src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+        }
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static KeySet * getKeySet(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    return it == sGroupKeys.end() ? nullptr : it->second;
+}
+
+static EmberAfStatus setKey(uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan & keyRoot, chip::ByteSpan & epochStartTime,
+                            uint8_t securityPolicy)
+{
+    if (keyRoot.size() != GROUP_KEY_ROOT_SIZE || epochStartTime.size() != GROUP_KEY_EPOCH_SIZE)
+    {
+        return EMBER_ZCL_STATUS_INVALID_VALUE;
+    }
+    else
+    {
+        KeySet * set = getKeySet(fabricIndex, keySetIndex);
+        if (set)
+        {
+            set->current = nextEpoch(set->current);
+        }
+        else
+        {
+            set                                                                  = new KeySet();
+            sGroupKeys[std::tuple<uint16_t, uint16_t>(fabricIndex, keySetIndex)] = set;
+        }
+        RootKey & key = set->epoch[set->current];
+        key.active    = true;
+        memcpy(key.value, keyRoot.data(), GROUP_KEY_ROOT_SIZE);
+        memcpy(key.startTime, epochStartTime.data(), GROUP_KEY_EPOCH_SIZE);
+        printKeys();
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+}
+
+static EmberAfStatus removeKey(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    emberAfGroupKeyManagementClusterPrint("removeKey(%u, %u)", fabricIndex, keySetIndex);
+
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    EmberAfStatus status                                            = EMBER_ZCL_STATUS_SUCCESS;
+
+    if (it == sGroupKeys.end())
+    {
+        // No key found
+        status = EMBER_ZCL_STATUS_NOT_FOUND;
+    }
+    else
+    {
+        // Existing key
+        delete it->second;
+        sGroupKeys.erase(it);
+    }
+
+    printKeys();
+    revokeKeys(fabricIndex, keySetIndex);
+    return status;
+}
+
+static EmberAfStatus removeAllKeys(uint16_t fabricIndex)
+{
+    emberAfGroupKeyManagementClusterPrint("removeAllKeys(%u)", fabricIndex);
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.begin();
+    while (it != sGroupKeys.end())
+    {
+        auto curr = it++;
+        if (std::get<0>(curr->first) == fabricIndex)
+        {
+            delete curr->second;
+            sGroupKeys.erase(curr);
+        }
+    }
+    printKeys();
+    revokeKeys(fabricIndex);
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static EmberAfStatus printStatus()
+{
+    emberAfGroupKeyManagementClusterPrint("GROUP STATES(%u)", sGroupStates.size());
+
+    std::set<std::tuple<uint16_t, GroupId, uint16_t>>::iterator it;
+    uint16_t fabricIndex = 0;
+    uint16_t groupId     = 0;
+    uint16_t keySet      = 0;
```

Comment:
These are only used in the loop. They should be hoisted down at the time of single use with declaration + assignment on one line, in the scope.